### PR TITLE
build(api, worker, ws, webhook): Use Nest.js SWC compiler plugin

### PR DESCRIPTION
**Pull request by rifont**
### What changed? Why was the change needed?

* Use SWC for Nest.js application builds
* Refactor env var management for all Nest.js apps to a single entry point
* Add missing `enum` reference to API properties to ensure enums are resolvable in Swagger CLI plugin
* Export all classes referenced by Swagger to ensure classes  are resolvable in Swagger CLI plugin

### Screenshots

_Successful Sanity test in local Studio, message is delivered_
image

_Apps with SWC (API, Worker, Webhook, Websocket)_
image

<details>
<summary><strong>Expand for optional sections</strong></summary>

### Special notes for your reviewer

</details>

### CODE REVIEW

File: apps/api/.swcrc
Comment (rifont) [inline apps/api/.swcrc:1]: Necessary when using Nest plugins.

File: apps/api/src/app/auth/auth.module.ts [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import { MiddlewareConsumer, Module, NestModule, Provider, RequestMethod } from '@nestjs/common';
 import { JwtModule } from '@nestjs/jwt';
 import { PassportModule } from '@nestjs/passport';
-import * as passport from 'passport';
+const passport = require('passport');
```
Comment (rifont) [inline apps/api/src/app/auth/auth.module.ts:4]: Broken with SWC.

File: .github/workflows/reusable-widget-e2e.yml [hunk 2]
```diff
@@ -74,7 +74,7 @@ jobs:
           NODE_ENV: 'test'
           TZ: 'UTC'
         run: |
-          cp apps/ws/src/.env.test apps/ws/dist/src/.env.test
+          cp apps/ws/src/.env.test apps/ws/dist/.env.test
```
Comment (rifont) [inline .github/workflows/reusable-widget-e2e.yml:77]: No more nested `src` in `dist` folder, it's flatter now.

File: apps/api/package.json [hunk 3]
```diff
@@ -64,8 +64,8 @@
     "compression": "^1.7.4",
     "cross-env": "^7.0.3",
     "date-fns": "^2.29.2",
-    "dotenv": "^8.2.0",
-    "envalid": "^6.0.1",
+    "dotenv": "^16.4.5",
+    "envalid": "^8.0.0",
```
Comment (rifont) [inline apps/api/package.json:63]: Upgrading to get latest features and fixes.

File: apps/api/package.json [hunk 4]
```diff
@@ -104,7 +104,7 @@
     "@types/chai": "^4.2.11",
     "@types/express": "4.17.17",
     "@types/mocha": "^10.0.2",
-    "@types/node": "^14.6.0",
+    "@types/node": "^20.14.10",
```
Comment (rifont) [inline apps/api/package.json:102]: Updated everywhere to Node v20

File: apps/api/src/app/invites/dtos/bulk-invite-members.dto.ts [hunk 5]
```diff
@@ -2,7 +2,7 @@ import { IBulkInviteRequestDto } from '@novu/shared';
 import { Type } from 'class-transformer';
 import { ArrayNotEmpty, IsArray, IsDefined, IsEmail, IsNotEmpty, ValidateNested } from 'class-validator';
 
-class EmailInvitee {
+export class EmailInvitee {
```
Comment (rifont) [inline apps/api/src/app/invites/dtos/bulk-invite-members.dto.ts:5]: We need to export all classes referenced by Swagger so it can compose the OpenAPI Spec (OAS) successfully.

File: apps/api/src/app/layouts/dtos/layout.dto.ts [hunk 6]
```diff
@@ -33,7 +33,9 @@ export class LayoutDto {
   @ApiProperty()
   description?: LayoutDescription;
 
-  @ApiProperty()
+  @ApiProperty({
+    enum: ChannelTypeEnum,
+  })
```
Comment (rifont) [inline apps/api/src/app/layouts/dtos/layout.dto.ts:38]: We need to reference the actual `enum` value everywhere so Swagger CLI has it for reference when composing the OAS

File: apps/api/src/app/organization/dtos/organization-response.dto.ts [hunk 7]
```diff
@@ -22,7 +22,7 @@ export class IPartnerConfigurationResponseDto {
   teamId: string;
 
   @ApiProperty({
-    enum: { ...PartnerTypeEnum },
+    enum: PartnerTypeEnum,
```
Comment (rifont) [inline apps/api/src/app/organization/dtos/organization-response.dto.ts:25]: Quite a few fixups like this to ensure the enum is available to Swagger

File: apps/api/src/app/shared/framework/swagger/swagger.controller.ts [hunk 8]
```diff
@@ -111,7 +112,8 @@ if (process.env.NOVU_ENTERPRISE === 'true') {
     { url: 'https://docs.novu.co/content-creation-design/translations' }
   );
 }
-export const setupSwagger = (app: INestApplication) => {
+export const setupSwagger = async (app: INestApplication) => {
+  await SwaggerModule.loadPluginMetadata(metadata);
```
Comment (rifont) [inline apps/api/src/app/shared/framework/swagger/swagger.controller.ts:116]: `metadata` is auto-generated by Swagger CLI during builds, it's necessary to provide back to the Swagger controller so it can create the Open API Spec successfully.

File: apps/api/src/config/env.validators.ts
Comment (rifont) [inline apps/api/src/config/env.validators.ts:1]: All `apps` now utilize this env-config pattern. We will define all env-vars in `env.validators` when they are needed, Typescript inference is used to populate a valid `process.env` typing everywhere.

File: packages/providers/src/lib/emailjs/emailjs.provider.ts [hunk 9]
```diff
@@ -6,9 +6,8 @@ import {
   ICheckIntegrationResponse,
   CheckIntegrationResponseEnum,
 } from '@novu/stateless';
-import { Message, SMTPClient, MessageAttachment } from 'emailjs';
 import { IEmailJsConfig } from './emailjs.config';
-import { MessageHeaders } from 'emailjs/smtp/message';
+import type { Message, SMTPClient, MessageAttachment } from 'emailjs';
```
Comment (rifont) [inline packages/providers/src/lib/emailjs/emailjs.provider.ts:10]: This provider shipped ESM only and wasn't compatible with SWC, so we shift to use a runtime require.

File: apps/inbound-mail/Dockerfile [hunk 10]
```diff
@@ -32,9 +32,9 @@ RUN --mount=type=secret,id=BULL_MQ_PRO_NPM_TOKEN,uid=1000 export BULL_MQ_PRO_NPM
 
 WORKDIR /usr/src/app/apps/inbound-mail
 
-RUN cp src/.env.test dist/src/.env.test
-RUN cp src/.env.development dist/src/.env.development
-RUN cp src/.env.production dist/src/.env.production
+RUN cp src/.env.test dist/.env.test
+RUN cp src/.env.development dist/.env.development
+RUN cp src/.env.production dist/.env.production
```
Comment (rifont) [inline apps/inbound-mail/Dockerfile]:
```suggestion
RUN cp src/.env.test dist/src/.env.test
RUN cp src/.env.development dist/src/.env.development
RUN cp src/.env.production dist/src/.env.production
```

File: apps/inbound-mail/Dockerfile [hunk 11]
```diff
@@ -69,4 +69,4 @@ RUN --mount=type=cache,id=pnpm-store-inbound-mail,target=/root/.pnpm-store\
  --reporter=silent
 
 WORKDIR /usr/src/app/apps/inbound-mail
-CMD [ "pm2-runtime", "dist/src/main.js" ]
+CMD [ "pm2-runtime", "dist/main.js" ]
```
Comment (rifont) [inline apps/inbound-mail/Dockerfile]:
```suggestion
CMD [ "pm2-runtime", "dist/src/main.js" ]
```

File: apps/inbound-mail/nest-cli.json
Comment (rifont) [inline apps/inbound-mail/nest-cli.json:1]: This isn't used.

File: apps/api/package.json [hunk 12]
```diff
@@ -9,25 +9,20 @@
     "prebuild": "rimraf dist",
     "build": "nest build",
     "format": "prettier --write \"src/**/*.ts\"",
-    "precommit": "lint-staged",
     "docker:build": "pnpm --silent --workspace-root pnpm-context -- apps/api/Dockerfile | BULL_MQ_PRO_NPM_TOKEN=${BULL_MQ_PRO_NPM_TOKEN} docker buildx build --load -t novu-api --secret id=BULL_MQ_PRO_NPM_TOKEN --build-arg PACKAGE_PATH=apps/api - $DOCKER_BUILD_ARGUMENTS",
     "docker:build:depot": "pnpm --silent --workspace-root pnpm-context -- apps/api/Dockerfile | depot build --build-arg PACKAGE_PATH=apps/api - -t novu-api --load",
     "start": "pnpm start:dev",
-    "start:dev": "cross-env TZ=UTC nest start --watch",
```
Comment (rifont) [inline apps/api/package.json:16]: No need for `TZ=UTC` anymore, it's handle in the env config defaults.

File: apps/api/src/app/auth/auth.module.ts (hunk 1 above)
Comment (SokratisVidros) [reply apps/api/src/app/auth/auth.module.ts:4]: Did we get to the bottom of this? Is it because passport is CJS only? I'd assume that a lot of packages fall under the same category.

File: libs/shared/src/config/processEnv.ts [hunk 13]
```diff
@@ -0,0 +1,29 @@
+const DEFAULT_ENV_FILE = '.env';
```
Comment (SokratisVidros) [inline libs/shared/src/config/processEnv.ts:1]: This is not part of this PR but is this custom .env resolution necessary? Can we just rely on dotenv defaults and avoid hardcoded per env files?

**Review by SokratisVidros — APPROVED** [review]
🐎

File: apps/api/src/app/auth/auth.module.ts (hunk 1 above)
Comment (rifont) [reply apps/api/src/app/auth/auth.module.ts:4]: Yes, it's because Passport is CJS only.

We need to make a decision about moving to pure ESM in the repo and a lot of these CJS-only packages will either:
1. need upgrades
2. replacements entirely for small things
3. Or, we wait for EMS compatibility, or port our own patch

File: libs/shared/src/config/processEnv.ts (hunk 13 above)
Comment (rifont) [reply libs/shared/src/config/processEnv.ts:1]: We probably can rely on dotenv defaults - my current thinking/goal is to get configuration aligned across the repo, then we'll look for commonalities and simplify.

Comment (SokratisVidros) [reply libs/shared/src/config/processEnv.ts:1]: We are aligned. It's a separate work/PR anyway.
